src/annotator/integrations/pdf2htmlEx.tsx

```diff
@@ -0,0 +1,551 @@
+
+    // Get the element that contains all of the PDF.js UI. This is typically
+    // `document.body`.
+    this._pdfContainer = pdf2htmlEXViewerApp.container?.parentElement ?? document.body;
+
+    // this._pdfMetadata = new PDFMetadata(pdf2htmlEXViewerApp);
+    this._htmlMeta = new HTMLMetadata();
+
+    this._observer = new MutationObserver(debounce(() => this._update(), 100));
+    this._observer.observe(this._pdfViewer.container, {
+      attributes: true,
+      attributeFilter: ['data-page-url'],
+      childList: true,
+      subtree: true,
+    });
+
+    this._reanchoringMaxWait = options.reanchoringMaxWait ?? 3000;
+    this._banner = new BannerController();
+    this._checkForSelectableText();
+    this._sideBySideActive = false;
+
+    // Hide annotation layer when the user is making a selection. The annotation
+    // layer appears above the invisible text layer and can interfere with text
+    // selection. See https://github.com/hypothesis/client/issues/1464.
+    this._updateAnnotationLayerVisibility = () => {
+      const selection = pdfWindow.getSelection()!;
+
+      // Add CSS class to indicate whether there is a selection. Annotation
+      // layers are then hidden by a CSS rule in `pdfjs-overrides.scss`.
+      this._pdfViewer.container.classList.toggle(
+        'is-selecting',
+        !selection.isCollapsed,
+      );
+    };
+
+    this._listeners = new ListenerCollection();
+    this._listeners.add(
+      document,
+      'selectionchange',
+      this._updateAnnotationLayerVisibility,
+    );
+
+    this._destroyed = false;
+  }
+
+  destroy() {
+    this.fitSideBySide({
+      // Dummy layout that will cause side-by-side mode to be undone.
+      expanded: false,
+      width: 0,
+      toolbarWidth: 0,
+      height: window.innerHeight,
+    });
+
+    this._listeners.removeAll();
+    this._pdfViewer.container.classList.remove('has-transparent-text-layer');
+    this._observer.disconnect();
+    this._banner.destroy();
+    this._destroyed = true;
+  }
+
+  /**
+   * Return the URL of the currently loaded PDF document.
+   */
+  async uri() {
+    return this._htmlMeta.uri();
+  }
+
+  /**
+   * Return the metadata (eg. title) for the currently loaded PDF document.
+   */
+  async getMetadata() {
+    return this._htmlMeta.getDocumentMetadata();
+  }
+
+  /**
+   * Display a banner at the top of the PDF viewer showing information about the
+   * current document.
+   */
+  showContentInfo(info: ContentInfoConfig) {
+    this._banner.setContentInfo(info);
+  }
+
+  /**
+   * Resolve serialized `selectors` from an annotation to a range.
+   */
+  anchor(root: HTMLElement, selectors: Selector[]): Promise<Range> {
+    const range = selectors.find(s => s.type === 'RangeSelector') as
+        | RangeSelector
+        | undefined;
+
+    let page = this._pdfViewer.pages.find(page => {
+      let xpath = xpathFromNode(page.page, root);
+      return range?.startContainer.startsWith(xpath)
```

Comment:
Couldn't find a better way than an xpath prefix comparison to find the page.

It also means that we're basically only relying on `RangeSelector` whereas the other selectors are deemed more performant or robust. Luckily, the DOM of the pdf2HtmlEx output is stable.